# Preparing data

To run the models, we need the following input: 
- X_train_res and y_train_res: resampled train data using SMOTE
- X_test and y_test 
- X_train_12: with only the measurements for stage 1 and 2 
- X_test_12: with only the measurements for stage 1 and 2 

Import libraries. 

In [45]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from collections import Counter

Standardized_complete_dataset is the dataframe I created with all the extracted features from TS fresh. 
It has 111 rows and 129 columns. 

In [46]:
merged_df = pd.read_csv('/Users/dionnespaltman/Desktop/V6/merged_df_11-06-2024.csv', sep=',')
merged_df.drop('Unnamed: 0', axis=1, inplace=True)

display(merged_df)

ID  Stage  Gender        Date  Location  Condition  Faintness  \
0      7    1.0       2  2019-10-28         0          2        1.0   
1     23    1.0       2  2020-08-01         1          2        1.0   
2     24    1.0       2  2020-01-22         1          2        1.0   
3     25    1.0       2  2020-05-02         1          2        1.0   
4     26    1.0       1  2020-06-02         2          1        2.0   
..   ...    ...     ...         ...       ...        ...        ...   
273  323    1.0       2  2022-10-26         3          3        1.0   
274  325    1.0       2  2022-07-11         3          3        1.0   
275  329    1.0       2  2022-11-28         3          3        1.0   
276  330    1.0       2  2022-11-30         0          3        1.0   
277  331    1.0       2  2022-11-30         0          3        1.0   

     Dizziness  Weakness  Lightheadedness  ...  AU26_r__standard_deviation  \
0          1.0       1.0              2.0  ...                   -0.585935   
1          2.0       2.0              1.0  ...                   -0.624259   
2          1.0       1.0              1.0  ...                    2.367560   
3          1.0       1.0              1.0  ...                    0.584439   
4          2.0       2.0              1.0  ...                   -0.517633   
..         ...       ...              ...  ...                         ...   
273        1.0       1.0              1.0  ...                   -0.804403   
274        1.0       1.0              1.0  ...                   -1.926084   
275        1.0       1.0              1.0  ...                   -0.266200   
276        1.0       1.0              1.0  ...                   -1.889172   
277        1.0       1.0              1.0  ...                    1.340280   

     AU26_r__maximum  AU26_r__mean  AU26_r__root_mean_square  \
0           0.425279     -0.786613                 -0.696839   
1           0.425279     -0.642748                 -0.651193   
2           0.425279      3.132948                  2.795771   
3           0.425279      0.394186                  0.499951   
4           0.425279     -0.951628                 -0.731221   
..               ...           ...                       ...   
273        -0.162826     -0.846163                 -0.845794   
274         0.425279     -1.669019                 -1.846330   
275         3.521479      0.610696                  0.169932   
276        -3.328215     -1.706598                 -1.845208   
277         0.425279     -1.779581                  0.208143   

     AU45_r__sum_values  AU45_r__variance  AU45_r__standard_deviation  \
0              0.520966          1.700927                    1.712210   
1             -0.504468          0.296375                    0.440450   
2              0.129283         -0.279627                   -0.201661   
3             -0.587993          0.094357                    0.226315   
4             -1.239117         -0.246984                   -0.162414   
..                  ...               ...                         ...   
273           -0.142084          0.642278                    0.784992   
274           -0.586609         -0.362405                   -0.303044   
275           -2.381393          4.280036                    3.507346   
276           -0.958863         -0.594948                   -0.603916   
277            0.696288          6.381711                    4.701091   

     AU45_r__maximum  AU45_r__mean  AU45_r__root_mean_square  
0           0.830151      1.741972                  1.745418  
1           0.760722      1.243173                  0.665506  
2           0.014352     -0.478731                 -0.279724  
3          -0.645230      0.514820                  0.302852  
4           0.639220     -1.034420                 -0.377156  
..               ...           ...                       ...  
273         0.864866      0.560518                  0.737070  
274         0.326786     -0.488608                 -0.360082  
275

In [47]:
print(list(merged_df.columns))

['ID', 'Stage', 'Gender', 'Date', 'Location', 'Condition', 'Faintness', 'Dizziness', 'Weakness', 'Lightheadedness', 'Fear', 'Tension', 'Stress', 'Physical_sum', 'Psychological_sum', 'Sum_VVR', 'Sum_12', 'Sum_456', 'Sum_4567', 'Sum_1', 'Sum_2', 'VVR_group', 'AU01_r__sum_values', 'AU01_r__variance', 'AU01_r__standard_deviation', 'AU01_r__maximum', 'AU01_r__mean', 'AU01_r__root_mean_square', 'AU02_r__sum_values', 'AU02_r__variance', 'AU02_r__standard_deviation', 'AU02_r__maximum', 'AU02_r__mean', 'AU02_r__root_mean_square', 'AU04_r__sum_values', 'AU04_r__variance', 'AU04_r__standard_deviation', 'AU04_r__maximum', 'AU04_r__mean', 'AU04_r__root_mean_square', 'AU05_r__sum_values', 'AU05_r__variance', 'AU05_r__standard_deviation', 'AU05_r__maximum', 'AU05_r__mean', 'AU05_r__root_mean_square', 'AU06_r__sum_values', 'AU06_r__variance', 'AU06_r__standard_deviation', 'AU06_r__maximum', 'AU06_r__mean', 'AU06_r__root_mean_square', 'AU07_r__sum_values', 'AU07_r__variance', 'AU07_r__standard_deviatio

In [48]:
# Make this copy to save the information easily without having to import it again 
copy_merged_df = merged_df
print(copy_merged_df.shape)

(278, 124)


In [ ]:
# columns_to_drop = ['Stage', 'Gender', 'Date', 'Location', 'Condition', 'Faintness',
#                     'Dizziness', 'Weakness', 'Lightheadedness', 'Fear', 'Tension', 'Stress', 'Physical_sum', 
#                     'Psychological_sum', 'Sum_VVR', 'Sum_12', 'Sum_456', 'Sum_4567']

columns_to_drop = ['Stage', 'Location', 'Faintness',
                    'Dizziness', 'Weakness', 'Lightheadedness', 'Fear', 'Tension', 'Stress', 'Physical_sum', 
                    'Psychological_sum', 'Sum_VVR', 'Sum_12', 'Sum_4567']

merged_df = merged_df.drop(columns_to_drop, axis=1)

In [53]:
print(list(merged_df.columns))

['ID', 'Gender', 'Date', 'Condition', 'Sum_456', 'Sum_1', 'Sum_2', 'VVR_group', 'AU01_r__sum_values', 'AU01_r__variance', 'AU01_r__standard_deviation', 'AU01_r__maximum', 'AU01_r__mean', 'AU01_r__root_mean_square', 'AU02_r__sum_values', 'AU02_r__variance', 'AU02_r__standard_deviation', 'AU02_r__maximum', 'AU02_r__mean', 'AU02_r__root_mean_square', 'AU04_r__sum_values', 'AU04_r__variance', 'AU04_r__standard_deviation', 'AU04_r__maximum', 'AU04_r__mean', 'AU04_r__root_mean_square', 'AU05_r__sum_values', 'AU05_r__variance', 'AU05_r__standard_deviation', 'AU05_r__maximum', 'AU05_r__mean', 'AU05_r__root_mean_square', 'AU06_r__sum_values', 'AU06_r__variance', 'AU06_r__standard_deviation', 'AU06_r__maximum', 'AU06_r__mean', 'AU06_r__root_mean_square', 'AU07_r__sum_values', 'AU07_r__variance', 'AU07_r__standard_deviation', 'AU07_r__maximum', 'AU07_r__mean', 'AU07_r__root_mean_square', 'AU09_r__sum_values', 'AU09_r__variance', 'AU09_r__standard_deviation', 'AU09_r__maximum', 'AU09_r__mean', 'AU

In [54]:
print(merged_df.shape)

(278, 110)


In [55]:
print(list(merged_df.columns))

['ID', 'Gender', 'Date', 'Condition', 'Sum_456', 'Sum_1', 'Sum_2', 'VVR_group', 'AU01_r__sum_values', 'AU01_r__variance', 'AU01_r__standard_deviation', 'AU01_r__maximum', 'AU01_r__mean', 'AU01_r__root_mean_square', 'AU02_r__sum_values', 'AU02_r__variance', 'AU02_r__standard_deviation', 'AU02_r__maximum', 'AU02_r__mean', 'AU02_r__root_mean_square', 'AU04_r__sum_values', 'AU04_r__variance', 'AU04_r__standard_deviation', 'AU04_r__maximum', 'AU04_r__mean', 'AU04_r__root_mean_square', 'AU05_r__sum_values', 'AU05_r__variance', 'AU05_r__standard_deviation', 'AU05_r__maximum', 'AU05_r__mean', 'AU05_r__root_mean_square', 'AU06_r__sum_values', 'AU06_r__variance', 'AU06_r__standard_deviation', 'AU06_r__maximum', 'AU06_r__mean', 'AU06_r__root_mean_square', 'AU07_r__sum_values', 'AU07_r__variance', 'AU07_r__standard_deviation', 'AU07_r__maximum', 'AU07_r__mean', 'AU07_r__root_mean_square', 'AU09_r__sum_values', 'AU09_r__variance', 'AU09_r__standard_deviation', 'AU09_r__maximum', 'AU09_r__mean', 'AU

These are the columns I use to predict, so all my features. I need these as a list to establish my featurizer. 
I have 102 features from TS fresh and then I added the two VVR measurements from stage 1 and 2. 

In [56]:
columns_ts_fresh = ['AU01_r__sum_values', 'AU01_r__variance', 'AU01_r__standard_deviation', 'AU01_r__maximum', 
                    'AU01_r__mean', 'AU01_r__root_mean_square', 'AU02_r__sum_values', 'AU02_r__variance', 
                    'AU02_r__standard_deviation', 'AU02_r__maximum', 'AU02_r__mean', 'AU02_r__root_mean_square', 
                    'AU04_r__sum_values', 'AU04_r__variance', 'AU04_r__standard_deviation', 'AU04_r__maximum', 
                    'AU04_r__mean', 'AU04_r__root_mean_square', 'AU05_r__sum_values', 'AU05_r__variance',
                    'AU05_r__standard_deviation', 'AU05_r__maximum', 'AU05_r__mean', 'AU05_r__root_mean_square', 
                    'AU06_r__sum_values', 'AU06_r__variance', 'AU06_r__standard_deviation', 'AU06_r__maximum', 
                    'AU06_r__mean', 'AU06_r__root_mean_square', 'AU07_r__sum_values', 'AU07_r__variance', 
                    'AU07_r__standard_deviation', 'AU07_r__maximum', 'AU07_r__mean', 'AU07_r__root_mean_square', 
                    'AU09_r__sum_values', 'AU09_r__variance', 'AU09_r__standard_deviation', 'AU09_r__maximum',
                    'AU09_r__mean', 'AU09_r__root_mean_square', 'AU10_r__sum_values', 'AU10_r__variance', 
                    'AU10_r__standard_deviation', 'AU10_r__maximum', 'AU10_r__mean', 'AU10_r__root_mean_square',
                    'AU12_r__sum_values', 'AU12_r__variance', 'AU12_r__standard_deviation', 'AU12_r__maximum', 
                    'AU12_r__mean', 'AU12_r__root_mean_square', 'AU14_r__sum_values', 'AU14_r__variance', 
                    'AU14_r__standard_deviation', 'AU14_r__maximum', 'AU14_r__mean', 'AU14_r__root_mean_square', 
                    'AU15_r__sum_values', 'AU15_r__variance', 'AU15_r__standard_deviation', 'AU15_r__maximum', 
                    'AU15_r__mean', 'AU15_r__root_mean_square', 'AU17_r__sum_values', 'AU17_r__variance', 
                    'AU17_r__standard_deviation', 'AU17_r__maximum', 'AU17_r__mean', 'AU17_r__root_mean_square', 
                    'AU20_r__sum_values', 'AU20_r__variance', 'AU20_r__standard_deviation', 'AU20_r__maximum', 
                    'AU20_r__mean', 'AU20_r__root_mean_square', 'AU23_r__sum_values', 'AU23_r__variance', 
                    'AU23_r__standard_deviation', 'AU23_r__maximum', 'AU23_r__mean', 'AU23_r__root_mean_square', 
                    'AU25_r__sum_values', 'AU25_r__variance', 'AU25_r__standard_deviation', 'AU25_r__maximum', 
                    'AU25_r__mean', 'AU25_r__root_mean_square', 'AU26_r__sum_values', 'AU26_r__variance', 
                    'AU26_r__standard_deviation', 'AU26_r__maximum', 'AU26_r__mean', 'AU26_r__root_mean_square', 
                    'AU45_r__sum_values', 'AU45_r__variance', 'AU45_r__standard_deviation', 'AU45_r__maximum', 
                    'AU45_r__mean', 'AU45_r__root_mean_square']

print(len(columns_ts_fresh))

102


# Train and test split
May  30, 2024: 
- In total, we have 104 participants.
- I started with a test size of 20%. Then there are 83 people in the train set and 21 in the test set. With a test size of 30#, there are 72 people in the train set and 32 in the test set. 
- Naturally, we stratify on VVR_group. 

June 4, 2024: 
- In total, we have 111 participants.
- I started with a test size of 20%. Then there are 88 people in the train set and 23 in the test set. With a test size of 30% there are 77 people in the train set and 34 in the test set. 
- Naturally, we stratify on VVR_group. 

June 11, 2024: 
- In total, we have 278 participants. 
- Test size of 20%: 222 in the train and 56 in the test set. 
- Test size of 30%: 184 in the train and 84 in the test set. 

In [57]:
train, test = train_test_split(merged_df, test_size=0.3, random_state=123, stratify=merged_df['VVR_group'])

print(train.shape)
print(test.shape)

(194, 110)
(84, 110)


May 30, 2024: Unfortunately, the test set is very small with only 7 people in the high VVR condition. 

June 3, 2024: The test set is very small with only 11 people in the high VVR condition. 

June 11, 2024: 56 people is the low VVR group and 28 in the high VVR group. 

In [19]:
columns_to_drop = [ 'ID', 'VVR_group'] 

X_test = test.drop(columns_to_drop, axis=1)
y_test = test['VVR_group']

# Print original class distribution
print('Original dataset shape %s' % Counter(y_test))

Original dataset shape Counter({0: 56, 1: 28})


In [20]:
display(X_test)

Gender        Date  Condition  Sum_456  Sum_1  Sum_2  AU01_r__sum_values  \
210       1  2022-01-26          3     34.0   11.0   11.0            0.692606   
258       1  2022-08-15          3     35.0   13.0   11.0           -0.499063   
141       1  2021-12-07          3     27.0    8.0    9.0           -0.286902   
21        2  2020-07-14          2     40.0   17.0   16.0           -0.334424   
244       1  2022-06-20          3     24.0    9.0    9.0           -0.886020   
..      ...         ...        ...      ...    ...    ...                 ...   
131       1  2021-06-28          2     27.0    9.0    9.0           -0.471048   
137       1  2021-06-07          3     25.0   10.0   10.0            2.344491   
109       2  2021-05-26          3     28.0    8.0    8.0            0.655974   
188       2  2021-11-25          3     29.0   16.0   16.0           -0.986948   
133       2  2021-06-29          3     36.0   12.0    9.0           -0.197909   

     AU01_r__variance  AU01_r__standard_deviation  AU01_r__maximum  ...  \
210          1.893534                    1.766233         0.293698  ...   
258         -0.220853                   -0.102822         0.090348  ...   
141          0.506673                    0.634509         0.293698  ...   
21          -0.968126                   -1.075476         0.293698  ...   
244         -0.628347                   -0.593900        -0.830077  ...   
..                ...                         ...              ...  ...   
131          0.292913                    0.431894         0.293698  ...   
137          2.207046                    1.990390         0.293698  ...   
109         -0.194846                   -0.073803         0.293698  ...   
188         -0.800269                   -0.827082        -1.675583  ...   
133         -0.623426                   -0.587491         0.036835  ...   

     AU26_r__standard_deviation  AU26_r__maximum  AU26_r__mean  \
210                    0.229617         0.425279      0.465267   
258                   -1.430830        -2.601732     -1.454398   
141                   -1.108111        -1.633089     -0.546063   
21                     0.258397         0.425279      0.364450   
244                    0.487098         0.425279      0.031160   
..                          ...              ...           ...   
131                   -0.664260         0.425279     -0.465377   
137                    0.929832         0.425279      1.399776   
109                   -0.604077         0.425279     -0.802144   
188                   -0.990906        -1.425522     -0.839147   
133                   -1.262881         0.200415     -1.119870   

     AU26_r__root_mean_square  AU45_r__sum_values  AU45_r__variance  \
210                  0.343728           -0.936825         -0.962588   
258                 -1.475502           -0.159630          0.251873   
141                 -0.855302           -1.324306         -1.006654   
21                   0.309274            1.277550          0.471538   
244                  0.278753           -1.145047         -0.955903   
..                        ...                 ...               ...   
131                 -0.586977           -0.586505         -0.541951   
137                  1.176559           -0.642753         -0.886359   
109                 -0.714106            0.575739         -0.602680   
188                 -0.943481           -1.491900         -1.022418   
133                 -1.225289            1.428425          1.137925   

     AU45_r__standard_deviation  AU45_r__maximum  AU45_r__mean  \
210                   -1.143071        -1.443672     -0.980686   
258                    0.394171        -1.391600      0.812650   
141                   -1.214641         0.465645     -0.996433   
21                     0.618123         0.864866      0.307900   
244                   -1.132368        -1.669319     -1.220803   
..                          ...              ...           ...   
131                   -

# Applying SMOTE
We apply SMOTE on the train data. The strategy is to make the minority as big as the majority class. 

In [21]:
from collections import Counter
from sklearn.datasets import make_classification
from imblearn.over_sampling import SMOTE

In [26]:
columns_to_drop = [ 'ID', 'VVR_group', 'Gender', 'Date', 'Condition'] 

X_train = train.drop(columns_to_drop, axis=1)
y_train = train['VVR_group']

display(X_train)

Sum_456  Sum_1  Sum_2  AU01_r__sum_values  AU01_r__variance  \
212     31.0   14.0   13.0            0.045838          0.099566   
139     36.0   12.0   11.0            0.855351          0.420324   
40      43.0   19.0   21.0           -0.220723          0.583830   
52      26.0   11.0   11.0           -1.883382         -1.549193   
245     24.0    8.0    8.0           -1.097556         -1.219151   
..       ...    ...    ...                 ...               ...   
135     27.0    9.0    9.0           -0.173562         -0.410987   
1       27.0   13.0   11.0           -1.308169         -0.495347   
261     28.0   10.0   10.0           -0.915057         -0.901335   
77      26.0   10.0    9.0            1.176694          1.173271   
233     34.0   10.0   11.0            0.215667         -0.479852   

     AU01_r__standard_deviation  AU01_r__maximum  AU01_r__mean  \
212                    0.239420         0.293698      0.007298   
139                    0.553853         0.293698      0.872000   
40                     0.705300         0.293698      0.952411   
52                    -2.252417         0.111753     -1.872637   
245                   -1.501183        -0.380567     -1.553547   
..                          ...              ...           ...   
135                   -0.322775         0.293698     -0.142622   
1                     -0.425249         0.293698     -0.703835   
261                   -0.973820        -0.690943     -0.790561   
77                     1.211837         0.293698      1.119853   
233                   -0.406180        -0.059488     -0.251809   

     AU01_r__root_mean_square  AU02_r__sum_values  ...  \
212                  0.171839            0.046312  ...   
139                  0.654390            0.058924  ...   
40                   0.786203           -0.246186  ...   
52                  -2.179367           -1.845858  ...   
245                 -1.540047           -1.073593  ...   
..                        ...                 ...  ...   
135                 -0.279186           -0.077088  ...   
1                   -0.513168           -1.253416  ...   
261                 -0.939593           -0.661718  ...   
77                   1.198375            0.419833  ...   
233                 -0.371667            0.190405  ...   

     AU26_r__standard_deviation  AU26_r__maximum  AU26_r__mean  \
212                   -0.350670         0.425279     -0.257224   
139                    0.546475         0.338793      0.701921   
40                     2.336406         0.425279      2.014164   
52                    -1.076953         0.425279     -1.316843   
245                   -1.746968        -2.792001     -1.613128   
..                          ...              ...           ...   
135                    0.313968         0.425279      0.443651   
1                     -0.624259         0.425279     -0.642748   
261                    0.217045         0.200415      0.549468   
77                    -0.590969        -0.370392     -0.656436   
233                    0.197423         0.425279      0.297062   

     AU26_r__root_mean_square  AU45_r__sum_values  AU45_r__variance  \
212                 -0.318448           -0.407479         -0.798105   
139                  0.627531            0.516766          0.532184   
40                   2.223629           -0.297024          0.751723   
52                  -1.213479           -1.115520          0.159606   
245                 -1.723628            0.043857         -0.264156   
..                        ...                 ...               ...   
135                  0.377551           -0.187736         -0.387676   
1                   -0.651193           -0.504468          0.296375   
261                  0.379275           -1.307713         -1.122401   
77                  -0.639399            1.276020          0.529910   
233                  0.243914            0.102001         -0.280416   

     AU45_r__standard_deviation  AU45_r__maximum  AU

In [27]:
# Print original class distribution
print('Original dataset shape %s' % Counter(y_train))

# Apply SMOTE to the training data with sampling strategy set to 'auto' (default)
sm = SMOTE(sampling_strategy='not majority', random_state=42, k_neighbors=5)
X_train_res, y_train_res = sm.fit_resample(X_train, y_train)

# Print resampled class distribution
print('Resampled dataset shape %s' % Counter(y_train_res))

Original dataset shape Counter({0: 130, 1: 64})
Resampled dataset shape Counter({0: 130, 1: 130})


In [29]:
display(X_train_res)
X_train_res.to_csv('/Users/dionnespaltman/Desktop/V6/X_train_resampled_12-06-2024.csv')

Sum_456      Sum_1      Sum_2  AU01_r__sum_values  AU01_r__variance  \
0    31.000000  14.000000  13.000000            0.045838          0.099566   
1    36.000000  12.000000  11.000000            0.855351          0.420324   
2    43.000000  19.000000  21.000000           -0.220723          0.583830   
3    26.000000  11.000000  11.000000           -1.883382         -1.549193   
4    24.000000   8.000000   8.000000           -1.097556         -1.219151   
..         ...        ...        ...                 ...               ...   
255  30.440290  11.186570   8.000000           -0.363042         -0.716607   
256  38.785118  12.000000  10.322323           -0.155563         -0.557816   
257  33.303289  10.921316   9.381973            0.292014          0.358228   
258  29.770239  15.577680  10.770239            0.649773          0.798889   
259  37.415635  12.792183  11.000000           -0.372075          0.058708   

     AU01_r__standard_deviation  AU01_r__maximum  AU01_r__mean  \
0                      0.239420         0.293698      0.007298   
1                      0.553853         0.293698      0.872000   
2                      0.705300         0.293698      0.952411   
3                     -2.252417         0.111753     -1.872637   
4                     -1.501183        -0.380567     -1.553547   
..                          ...              ...           ...   
255                   -0.716452        -0.110127     -0.871241   
256                   -0.503356         0.289098     -0.295854   
257                    0.401179         0.293698      0.362417   
258                    0.882572         0.293698      0.898829   
259                    0.180254         0.185825      0.314582   

     AU01_r__root_mean_square  AU02_r__sum_values  ...  \
0                    0.171839            0.046312  ...   
1                    0.654390            0.058924  ...   
2                    0.786203           -0.246186  ...   
3                   -2.179367           -1.845858  ...   
4                   -1.540047           -1.073593  ...   
..                        ...                 ...  ...   
255                 -0.773932            0.181110  ...   
256                 -0.452584           -0.363124  ...   
257                  0.391054            0.869975  ...   
258                  0.896147           -0.337743  ...   
259                  0.218454           -0.099846  ...   

     AU26_r__standard_deviation  AU26_r__maximum  AU26_r__mean  \
0                     -0.350670         0.425279     -0.257224   
1                      0.546475         0.338793      0.701921   
2                      2.336406         0.425279      2.014164   
3                     -1.076953         0.425279     -1.316843   
4                     -1.746968        -2.792001     -1.613128   
..                          ...              ...           ...   
255                   -0.298781         0.425279     -0.597373   
256                   -0.650420         0.375101     -0.904794   
257                   -0.045150         0.425279      0.451452   
258                   -0.048664         0.368656     -0.231451   
259                    1.904184         0.698473      1.549799   

     AU26_r__root_mean_square  AU45_r__sum_values  AU45_r__variance  \
0                   -0.318448           -0.407479         -0.798105   
1                    0.627531            0.516766          0.532184   
2                    2.223629           -0.297024          0.751723   
3                   -1.213479           -1.115520          0.159606   
4                   -1.723628            0.043857         -0.264156   
..                        ...                 ...               ...   
255                 -0.449204            0.632487         -0.173541   
256                 -0.786691            0.045321         -0.416551   
257                  0.196125           -0.873549         -1.066030   
258                 -0.139576           -0.415063         -0.521501   
259   

# Adding class weights

I will add class weights to my models, because of my inbalanced data set. 
https://medium.com/@ravi.abhinav4/improving-class-imbalance-with-class-weights-in-machine-learning-af072fdd4aa4 

May 30, 2024: {0: 0.6428571428571429, 1: 2.25}

June 3, 2024: {0: 0.7549019607843137, 1: 1.4807692307692308}

June 11, 2024:  {0: 0.7461538461538462, 1: 1.515625}

In [30]:
import numpy as np

def calculate_class_weights(y):
    unique_classes, class_counts = np.unique(y, return_counts=True)
    total_samples = len(y)
    class_weights = {}

    for class_label, class_count in zip(unique_classes, class_counts):
        class_weight = total_samples / (2.0 * class_count)
        class_weights[class_label] = class_weight

    return class_weights

# Assuming 'y' contains the class labels (0s and 1s) for the binary classification problem
class_weights = calculate_class_weights(y_train)
print("Class weights:", class_weights)


Class weights: {0: 0.7461538461538462, 1: 1.515625}


# Save X_train_res, y_train_res, X_test, y_test

In [31]:
display(X_train_res)

Sum_456      Sum_1      Sum_2  AU01_r__sum_values  AU01_r__variance  \
0    31.000000  14.000000  13.000000            0.045838          0.099566   
1    36.000000  12.000000  11.000000            0.855351          0.420324   
2    43.000000  19.000000  21.000000           -0.220723          0.583830   
3    26.000000  11.000000  11.000000           -1.883382         -1.549193   
4    24.000000   8.000000   8.000000           -1.097556         -1.219151   
..         ...        ...        ...                 ...               ...   
255  30.440290  11.186570   8.000000           -0.363042         -0.716607   
256  38.785118  12.000000  10.322323           -0.155563         -0.557816   
257  33.303289  10.921316   9.381973            0.292014          0.358228   
258  29.770239  15.577680  10.770239            0.649773          0.798889   
259  37.415635  12.792183  11.000000           -0.372075          0.058708   

     AU01_r__standard_deviation  AU01_r__maximum  AU01_r__mean  \
0                      0.239420         0.293698      0.007298   
1                      0.553853         0.293698      0.872000   
2                      0.705300         0.293698      0.952411   
3                     -2.252417         0.111753     -1.872637   
4                     -1.501183        -0.380567     -1.553547   
..                          ...              ...           ...   
255                   -0.716452        -0.110127     -0.871241   
256                   -0.503356         0.289098     -0.295854   
257                    0.401179         0.293698      0.362417   
258                    0.882572         0.293698      0.898829   
259                    0.180254         0.185825      0.314582   

     AU01_r__root_mean_square  AU02_r__sum_values  ...  \
0                    0.171839            0.046312  ...   
1                    0.654390            0.058924  ...   
2                    0.786203           -0.246186  ...   
3                   -2.179367           -1.845858  ...   
4                   -1.540047           -1.073593  ...   
..                        ...                 ...  ...   
255                 -0.773932            0.181110  ...   
256                 -0.452584           -0.363124  ...   
257                  0.391054            0.869975  ...   
258                  0.896147           -0.337743  ...   
259                  0.218454           -0.099846  ...   

     AU26_r__standard_deviation  AU26_r__maximum  AU26_r__mean  \
0                     -0.350670         0.425279     -0.257224   
1                      0.546475         0.338793      0.701921   
2                      2.336406         0.425279      2.014164   
3                     -1.076953         0.425279     -1.316843   
4                     -1.746968        -2.792001     -1.613128   
..                          ...              ...           ...   
255                   -0.298781         0.425279     -0.597373   
256                   -0.650420         0.375101     -0.904794   
257                   -0.045150         0.425279      0.451452   
258                   -0.048664         0.368656     -0.231451   
259                    1.904184         0.698473      1.549799   

     AU26_r__root_mean_square  AU45_r__sum_values  AU45_r__variance  \
0                   -0.318448           -0.407479         -0.798105   
1                    0.627531            0.516766          0.532184   
2                    2.223629           -0.297024          0.751723   
3                   -1.213479           -1.115520          0.159606   
4                   -1.723628            0.043857         -0.264156   
..                        ...                 ...               ...   
255                 -0.449204            0.632487         -0.173541   
256                 -0.786691            0.045321         -0.416551   
257                  0.196125           -0.873549         -1.066030   
258                 -0.139576           -0.415063         -0.521501   
259   

In [33]:
X_train_res.to_csv('/Users/dionnespaltman/Desktop/V6/12-06-2024/X_train_resampled.csv')
y_train_res.to_csv('/Users/dionnespaltman/Desktop/V6/12-06-2024/y_train_resampled.csv')
X_test.to_csv('/Users/dionnespaltman/Desktop/V6/12-06-2024/X_test.csv')
y_test.to_csv('/Users/dionnespaltman/Desktop/V6/12-06-2024/y_test.csv')

# Data distibution: old data vs new data 

In [4]:
X_train_june = pd.read_csv('/Users/dionnespaltman/Desktop/V6/data_04-06-2024/X_train_resampled.csv')
y_train_june = pd.read_csv('/Users/dionnespaltman/Desktop/V6/data_04-06-2024/y_train_resampled.csv')
X_test_june = pd.read_csv('/Users/dionnespaltman/Desktop/V6/data_04-06-2024/X_test.csv')
y_test_june = pd.read_csv('/Users/dionnespaltman/Desktop/V6/data_04-06-2024/y_test.csv')

In [5]:
X_train_may = pd.read_csv('/Users/dionnespaltman/Desktop/V4/data/X_train_res.csv')
y_train_may = pd.read_csv('/Users/dionnespaltman/Desktop/V4/data/y_train_res.csv')
X_test_may = pd.read_csv('/Users/dionnespaltman/Desktop/V4/data/X_test.csv')
y_test_may = pd.read_csv('/Users/dionnespaltman/Desktop/V4/data/y_test.csv')

In [8]:
counts_train_may = y_train_may['VVR_group'].value_counts()
print(counts_train_may)

1    56
0    56
Name: VVR_group, dtype: int64


In [9]:
counts_train_june = y_train_june['VVR_group'].value_counts()
print(counts_train_june)

1    51
0    51
Name: VVR_group, dtype: int64


In [6]:
counts_test_may = y_test_may['VVR_group'].value_counts()
print(counts_test_may)


0    25
1     7
Name: VVR_group, dtype: int64


In [7]:
counts_test_june = y_test_june['VVR_group'].value_counts()
print(counts_test_june)

0    23
1    11
Name: VVR_group, dtype: int64


# Complete dataset 

In [34]:
merged_df = pd.read_csv('/Users/dionnespaltman/Desktop/V6/merged_df_11-06-2024.csv')

In [35]:
display(merged_df)

Unnamed: 0   ID  Stage  Gender        Date  Location  Condition  \
0             0    7    1.0       2  2019-10-28         0          2   
1             1   23    1.0       2  2020-08-01         1          2   
2             2   24    1.0       2  2020-01-22         1          2   
3             3   25    1.0       2  2020-05-02         1          2   
4             4   26    1.0       1  2020-06-02         2          1   
..          ...  ...    ...     ...         ...       ...        ...   
273         273  323    1.0       2  2022-10-26         3          3   
274         274  325    1.0       2  2022-07-11         3          3   
275         275  329    1.0       2  2022-11-28         3          3   
276         276  330    1.0       2  2022-11-30         0          3   
277         277  331    1.0       2  2022-11-30         0          3   

     Faintness  Dizziness  Weakness  ...  AU26_r__standard_deviation  \
0          1.0        1.0       1.0  ...                   -0.585935   
1          1.0        2.0       2.0  ...                   -0.624259   
2          1.0        1.0       1.0  ...                    2.367560   
3          1.0        1.0       1.0  ...                    0.584439   
4          2.0        2.0       2.0  ...                   -0.517633   
..         ...        ...       ...  ...                         ...   
273        1.0        1.0       1.0  ...                   -0.804403   
274        1.0        1.0       1.0  ...                   -1.926084   
275        1.0        1.0       1.0  ...                   -0.266200   
276        1.0        1.0       1.0  ...                   -1.889172   
277        1.0        1.0       1.0  ...                    1.340280   

     AU26_r__maximum  AU26_r__mean  AU26_r__root_mean_square  \
0           0.425279     -0.786613                 -0.696839   
1           0.425279     -0.642748                 -0.651193   
2           0.425279      3.132948                  2.795771   
3           0.425279      0.394186                  0.499951   
4           0.425279     -0.951628                 -0.731221   
..               ...           ...                       ...   
273        -0.162826     -0.846163                 -0.845794   
274         0.425279     -1.669019                 -1.846330   
275         3.521479      0.610696                  0.169932   
276        -3.328215     -1.706598                 -1.845208   
277         0.425279     -1.779581                  0.208143   

     AU45_r__sum_values  AU45_r__variance  AU45_r__standard_deviation  \
0              0.520966          1.700927                    1.712210   
1             -0.504468          0.296375                    0.440450   
2              0.129283         -0.279627                   -0.201661   
3             -0.587993          0.094357                    0.226315   
4             -1.239117         -0.246984                   -0.162414   
..                  ...               ...                         ...   
273           -0.142084          0.642278                    0.784992   
274           -0.586609         -0.362405                   -0.303044   
275           -2.381393          4.280036                    3.507346   
276           -0.958863         -0.594948                   -0.603916   
277            0.696288          6.381711                    4.701091   

     AU45_r__maximum  AU45_r__mean  AU45_r__root_mean_square  
0           0.830151      1.741972                  1.745418  
1           0.760722      1.243173                  0.665506  
2           0.014352     -0.478731                 -0.279724  
3          -0.645230      0.514820                  0.302852  
4           0.639220     -1.034420                 -0.377156  
..               ...           ...                       ...  
273         0.864866      0.560518                  0.737070  
274         0.326786     -0.488608                 -0.360082  
275         0.517718      2.061400                  3.218211  

In [36]:
counts = merged_df['VVR_group'].value_counts()
print(counts)

0    186
1     92
Name: VVR_group, dtype: int64
